### DISCUSSION TITLE
Select Instances by region

### DISCUSSION THREAD

**Discussion by DanBThom** [gdsfactory/kfactory#523] [General]
Thank you for KFactory, the fill_tiled function is quick and useful.  I would like to fill_tiled a polygon, and then use another polygon (which is a subset of the original polygon) or region to select the portion of tiled instances that fall within this polygon.  The purpose of such is to change the pointer/instance to a different cell, so that the grid is constant across the larger polygon, but the geometry can change.   This would have an application of augmented reality waveguides.

**Comment by sebastian-goeldi** [comment]
Hello,

Let me think a bit on how to do this. It is definitely possible, but might needs to have some pieces exposed. 


But first let me ask a few guestions:

- Are you trying to fill a user generated polygon or are you trying to fill an extracted area?
- Is the polygon very complex or rather simple (think rectangle vs patches of random polygons to fill)

If it's rather a simple polygon and is user generated, you should probably directly use `kf.kdb.Region.fill_region` or `kf.kdb.Region.fill_region_multi`.

Otherwise, let me know and I might be able to expose the right parameter to enable user setting (it will most likely be the origin, but we need to check that `multi` is disabled)

**Comment by DanBThom** [comment]
Hi Sebastian, thanks for your response.
1) User generated polygons.  So that a designer can modify the polygon for a different design.
2) Complex polygon. 
Effectively there would be a large polygon (complex) that would be subdivided into smaller polygon areas (also complex).  Each subdivided area would point to a unique cell to have the fill ratio be the same in the subdivided area.  The reason to change an instance is that the pitch over the large polygon needs to remain constant.  Otherwise I would fill each subdivided area with the fill; but because of the constant pitch requirement I envision it would be difficult to line up the fills of each subdivided polygon to be on the same pitch.

**Reply by sebastian-goeldi** [reply]
I see. So, the global pitch should already be maintained in the case of `multi=False` (it might not fill all the cases it should as it seems I forgot to actually use that case, though I am not 100% it's necessary). But therefore, it should not be necessary to have a larger polygon, it's just a set origin. If you were to fill multiple cells and have a certain instantiation against each other, it would be a bit more complex.

It can currently not be globally configured but I can very easily expose it.

I will confirm tomorrow morning and patch it in. The necessary parameter to be set is `origin` in the `FillTiledOperator` class in `utils/fill.py`

**Reply by DanBThom** [reply]
Thanks for looking into this: here is example code with comments in capitals in the section in question:
import kfactory as kf
from kfactory.utils.fill import fill_tiled

if __name__ == "__main__":
    #create two polygon sets of points.  One is a subset of the other.
    points = [(-14899.999999999996, -13000.000000000002),
     (-14699.999999999996, -16500.0),
     (-15999.999999999996, -19200.000000000004),
     (-18499.999999999996, -21200.0),
     (-20999.999999999996, -21700.000000000004),
     (-23500.0, -21000.000000000007),
     (-25799.999999999996, -19000.000000000004),
     (-35399.99999999999, 2000.0000000000005),
     (-30199.999999999996, 5300.000000000001),
     (-14899.999999999996, -13000.000000000002)]
    points_sub_polygon = [(-14899.999999999996, -13000.000000000002),
     (-14699.999999999996, -16500.0),
     (-15999.999999999996, -19200.000000000004),
     (-18499.999999999996, -21200.0),
     (-20999.999999999996, -21700.000000000004)]
    #allow for grid to be on an angle, user defined.
    angle = 25
   
    #create cells to fill
    c = kf.KCell()
    c_outer1 = kf.KCell()
    c_outer2 = kf.KCell()
    
    '''
    Create instances
    '''
    c_inst_rotate = c_outer1.create_inst(c)
    c_inst_top = c_outer2.create_inst(c_outer1)
    '''
    rotate the instance prior to tiling
    '''
    c_inst_rotate.drotate(angle)
    '''
    create shapes to receive tiling and put them in the lowest level cell "c"
    points_sub_polygon is a subset of points of points
    '''
    poly1  = c.shapes(kf.kcl.layer(1, 0)).insert(
        kf.kdb.DPolygon(points)
    )
    poly2 = c.shapes(kf.kcl.layer(1, 1)).insert(
        kf.kdb.DPolygon(points_sub_polygon)
    )
    #create cell to fill with.  Couple of ellipses.  1 to populate, the other to replace.
    fc1 = kf.KCell()
    fc2 = kf.KCell()
    fc1.shapes(fc1.kcl.layer(2, 0)).insert(kf.kdb.DPolygon.ellipse(kf.kdb.DBox(20, 40),10))
    fc2.shapes(fc2.kcl.layer(2, 0)).insert(kf.kdb.DPolygon.ellipse(kf.kdb.DBox(30, 30),10))
    
    #copied Sebastien's code on the gdsfactory Geometry page
    fill_tiled(
        c_outer1,
        fc1,
        [(kf.kdb.LayerInfo(1, 0), 0)],
        exclude_layers=[
            (kf.kdb.LayerInfo(10, 0), 100),
            (kf.kdb.LayerInfo(2, 0), 0),
            (kf.kdb.LayerInfo(3, 0), 0),
        ],
        x_space=5,
        y_space=5,
    )
    '''
    DO SOMETHING TO REPLACE THE INSTANCES OF THE fc1 SHAPES WITH fc2 SHAPESIN THE 
    '''
    #Rotate the intance back, so that the grid itself is at an angle    
    angle  = -angle
    c_inst_top.drotate(angle)
   
    
    #output to gds
    gdspath = "mylayout.gds"

**Reply by DanBThom** [reply]
Thanks Sebastien, took me a while to understand and implement your comments. Yes the multi =false fixes the pitch variation.  The only issue I see is the joining point between polygons 
<img width="269" alt="pic for tiling" src="https://github.com/user-attachments/assets/68ee0793-a552-4a20-8961-07b6ebc07152">

**Reply by DanBThom** [reply]
Or rather, multi = true fixes the situation.
I can't seem to get 
tile_border = (x,y)
or
tile_size = (x,y) 
to affect the layout.  I thought maybe I could alter the tile_border values to allow for polygon intersection with the tiled cell to place the cell.

**Reply by sebastian-goeldi** [reply]
I am not sure I fully understand. Are those two different cells? If so, there is no way to fill them at their cell level. If it is a global fill, then it should already take care of it (and it wpuld be a bug if it didn't)

**Reply by DanBThom** [reply]
The tiling only occurs on one cell to ensure the origin remains the same. As you can see, there is a gap on one period between two fill locations.  I would like to fill that 1 period gap, as optically (for AR waveguides) this would mean an optical flare (MTF degradation):

mport kfactory as kf
from kfactory.utils.fill import fill_tiled
import gdsfactory as gf



if __name__ == "__main__":
    '''
    Steps:
        1) create polygon vertices and sub polygon vertices (future version to accept design files)
        2) place sub polygon vertices in an iterable array
        3) for each sub polygon vertices create polygon on a new layer
        4) Creation of fill cell with rectangles of the pitch (x and y) in the fill cell
        5) fill the overall polygon with tiled arrays, iteratively modifying which cell is placed in which sub polygon using the kfactory layer selection
        6) add to fill cell the correct ellipse
    '''
    ''' 
    create two polygon sets of points. One is a subset of the other.
    '''
    points = [(-14899.999999999996, -13000.000000000002),
     (-14699.999999999996, -16500.0),
     (-15999.999999999996, -19200.000000000004),
     (-18499.999999999996, -21200.0),
     (-20999.999999999996, -21700.000000000004),
     (-23500.0, -21000.000000000007),
     (-25799.999999999996, -19000.000000000004),
     (-35399.99999999999, 2000.0000000000005),
     (-30199.999999999996, 5300.000000000001),
     (-14899.999999999996, -13000.000000000002)]
    points_sub1 = [(-14899.999999999996, -13000.000000000002),
     (-14699.999999999996, -16500.0),
     (-20999.999999999996, -21700.000000000004),
     (-23500.0, -21000.000000000007)]
    points_sub2 = [(-15999.999999999996, -19200.000000000004),
    (-18499.999999999996, -21200.0),
    (-20999.999999999996, -21700.000000000004),
    (-14699.999999999996, -16500.0)]
    points_sub3 = [(-23500.0, -21000.000000000007),
    (-25799.999999999996, -19000.000000000004),
    (-35399.99999999999, 2000.0000000000005),
    (-30199.999999999996, 5300.000000000001),
    (-14899.999999999996, -13000.000000000002)]
    sub_polygon_array=[points_sub1, points_sub2, points_sub3]
    #allow for grid to be on an angle, user defined.
    angle = 25  #example angle, future design file will specify
    ellipse_rotation = 10  #example anlge, future design file will specify
    pitchx = 10000  #10um  10000nm
    pitchy = 10000
    layer_num_pitch = 11  #create box in cell at the pitch with this GDS layer
    #create canvas
  
    c = kf.KCell()            #the original polygon outline is in this cell and the sub polygons of the outline are placed in this cell
    c_outer1 = kf.KCell()     #the rotated polygon outline is placed in this cell, rotated.  Also the 2D grating material is placed on this cell
    c_outer2 = kf.KCell()     #The cel  Couter1 is rotated back, along with the 2D grating material 
    
    '''
    Create instances
    c is in c_outer_1
    c_outer1 is in c_outer2
    '''
    c_inst_rotate = c_outer1.create_inst(c)  
    c_inst_top = c_outer2.create_inst(c_outer1)
    '''
    rotate the instance prior to tiling
    '''
    c_inst_rotate.drotate(angle)
    '''
    create shapes to receive tiling and put them in the lowest level cell "c"
    points_sub_polygon is a subset of points of points
    Place each sub polygon on a different layer (1,x).  this is needed for the tiling script
    '''
    #create the outside boundary of a polygon.  All other polygons will be inside this boundary
    poly_outside  = c.shapes(kf.kcl.layer(1, 0)).insert(kf.kdb.DPolygon(points))  
    '''
    ellipse creation on a layer KCell.kcl.layer(x,y)
    insert a shape
    kfactory.kdb.DPolygon.ellipse(kfactory.kdb.DBox(major axis length in um, minor axis length in um), number of vertices for the ellipse)
    '''
    '''create polygons for sub areas'''
    layer_exclusion_list=[]
    for index in enumerate(sub_polygon_array):
        #print(index[0])
        poly = c.shapes(kf.kcl.layer(1, int(index[0]+1))).insert(kf.kdb.DPolygon(sub_polygon_array[index[0]]))
        layer_exclusion_list.append((kf.kdb.LayerInfo(1,int(index[0]+1)),0))
    fill_cells =[]
    fill_cells_ellipse =[]
    print ('my layer exclusions list')
    print (layer_exclusion_list)
    for index in enumerate(sub_polygon_array):
        #print(index[0])
        #poly = c.shapes(kf.kcl.layer(1, int(index[0+1]))).insert(kf.kdb.DPolygon(sub_polygon_array[index[0]]))
        fc = kf.KCell()                                 #declare cell to be tiled (fc)
        fc_ellipse = kf.KCell()                         #declare cell inside of cell to be tiled (fc_ellipse)
        fc_ellipse_inst =fc.create_inst(fc_ellipse)     #create instance of fc_ellipse in fc
        fill_cells.append(fc)
        fill_cells_ellipse.append(fc_ellipse)
        fc.shapes(fc.kcl.layer(layer_num_pitch, 0)).insert(kf.kdb.Box(-pitchx*0.5, -pitchy*0.5, pitchx*0.5, pitchy*0.5))
        layer_exclusion_list_temp = layer_exclusion_list[:index[0]] + layer_exclusion_list[index[0] + 1:]
        print ('my layer exclusions list temp')
        print (layer_exclusion_list_temp)
        #always fill the full grating area polygon.  The sub areas are in the exclusion list.  This allows the tiling to remain on the same origin
        fill_tiled(
            c = c_outer1,
            fill_cell = fc,
            fill_layers = [(kf.kdb.LayerInfo(1, 0), 0)],
            exclude_layers =layer_exclusion_list_temp, 
            x_space=0,
            y_space=0,
            multi= True,
            tile_border=(1000,1000)
        )
        fc_ellipse.shapes(fc_ellipse.kcl.layer(2, 0)).insert(kf.kdb.DPolygon.ellipse(kf.kdb.DBox(0.3, 0.2),10))
        fc_ellipse_inst.drotate(ellipse_rotation)       #rotate the cell fc_ellipse
            
      
    '''
    At this point all gratings should have tiled cells which are arrays of the FC (fill cell) cells within the larger grating.
    Fill_tiled always calls the outer polygon, but cycles through layers (1,x) for each sub polygon
    Exclusion list adds previos sub polygons already tiled.
    '''
    #Rotate the intance back, so that the grid itself is at an angle    
    angle  = -angle
    c_inst_top.drotate(angle)
        
    #output to gds
    gdspath = "mylayout1.gds"
    c_outer2.write(gdspath)

**Reply by DanBThom** [reply]
Thanks Sebastien for all your help.  End result of the work around is to follow Sebastien's use of 
1. setting multi to false when calling tiling
Additional implementation for full tiling
2. Use shapely to shrink polygons to shrink the exclusion zone by the pitch
3. Place tiling on large polygon that encompasses all areas with exclusions by shrunk polygons, and a layer that sets the ptich.
<img width="384" alt="pic for tiling2" src="https://github.com/user-attachments/assets/9687066d-3907-4a0c-a5aa-9752fe56ae0e">

**Reply by sebastian-goeldi** [reply]
So, I am not entirely following the logic of your fill.

But what I can say is. On the top level there should only be one fill call. If it's multiple fill layers you want (or multiple exclude), just pass all of them to the fill call

But it will merge all the fill layers & regions and then substract any exclude layers & region.

I don't fully get why you need three fill calls.

Also shapely should not be necessary. You should use `kdb.Region.size` or `kdb.Region.minkowski_sum` depending on your use cases (these are all in dbu, so be aware of that, they are not in um)

Additionally, if you need to size any of those layers, that's the second argument to any exclude/include tuple.

**Reply by DanBThom** [reply]
I'll look into the kdb.Region.size and kdb.Region.minkowski_sum, as well as the exclude/include tuple, thanks for pointing these out.  As for the 3 fill calls, I'm filling with different fill cells (you can see the ellipsis are different in different regions).  This is because I am using the tiling to put down features that will vary by region for different light extraction effects.

**Reply by sebastian-goeldi** [reply]
Ah, now I fully understand. Ok, in this case you should add the bounding boxes of the previous cells as excludes (ideally you have a layer, but it is possible without layer and just picking the instances) but also add previous polygon areas as fill.

Otherwise as you have seen, it will not fill border areas (because otherwise it would place fill outside the allowed region)
